tests/mocha/test_helpers.js

```diff
@@ -81,6 +84,86 @@ function createEventsFireStub() {
   return stub;
 }
 
+/**
+ * Shared setup method that sets up fake timer for clock so that pending
+ * setTimeout calls can be cleared in test teardown along with other common
+ * stubs. Should be called in setup of outermost suite using
+ * sharedTestSetup.call(this).
+ * @param {Object<string, boolean>} options Options to enable/disable setup
+ *    of certain stubs.
+ */
+function sharedTestSetup(options = {}) {
+  this.sharedSetupCalled_ = true;
+  // Sandbox created for greater control when certain stubs are cleared.
+  this.sharedSetupSandbox_ = sinon.createSandbox();
+  this.clock = this.sharedSetupSandbox_.useFakeTimers();
+  if (options['fireEventsNow'] === undefined || options['fireEventsNow']) {
+    // Stubs event firing unless passed option "fireEventsNow: false"
+    this.eventsFireStub = createEventsFireStubFireImmediately_(this.clock);
+  }
+}
+
+/**
+ * Shared cleanup method that clears up pending setTimeout calls, disposes of
+ * workspace, and resets global variables. Should be called in setup of
+ * outermost suite using sharedTestTeardown.call(this).
+ */
+function sharedTestTeardown() {
+  if (!this.sharedSetupCalled_) {
+    console.error('"' + this.currentTest.fullTitle() +
+        '" did not call sharedTestSetup');
+  }
+
+  try {
+    if (this.workspace) {
+      workspaceTeardown.call(this, this.workspace);
+      this.workspace = null;
+    } else {
+      this.clock.runAll();  // Run all queued setTimeout calls.
+    }
+  } catch (e) {
+    console.error(this.currentTest.fullTitle() + '\n', e);
+  } finally {
+    // Clear Blockly.Event state.
+    Blockly.Events.setGroup(false);
+    Blockly.Events.disabled_ = 0;
+    if (Blockly.Events.FIRE_QUEUE_.length) {
+      // If this happens, it may mean that some previous test is missing cleanup
+      // (i.e. a previous test added an event to the queue on a timeout that
+      // did not use a stubbed clock).
+      Blockly.Events.FIRE_QUEUE_.length = 0;
+      console.warn(this.currentTest.fullTitle() +
+          '" needed cleanup of Blockly.Events.FIRE_QUEUE_. This may indicate ' +
+          'leakage from an earlier test');
```

Comment:
This does currently cause warnings on the console, but I think they are valuable as it does indicate an actual issue that we should fix.